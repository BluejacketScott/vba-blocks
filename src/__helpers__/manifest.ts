import { dir as FIXTURES } from '../../tests/__fixtures__';
import { Manifest } from '../manifest';
import { isPathDependency } from '../manifest/dependency';
import { relative } from '../utils/path';

export function normalizeManifest(manifest: Manifest, relativeTo: string = FIXTURES): Manifest {
  for (const src of manifest.src) {
    src.path = normalizePath(src.path, relativeTo);

    if (src.binary) {
      src.binary = normalizePath(src.binary, relativeTo);
    }
  }
  for (const dependency of manifest.dependencies) {
    if (isPathDependency(dependency)) {
      dependency.path = normalizePath(dependency.path, relativeTo);
    }
  }
  for (const src of manifest.devSrc) {
    src.path = normalizePath(src.path, relativeTo);

    if (src.binary) {
      src.binary = normalizePath(src.binary, relativeTo);
    }
  }
  for (const dependency of manifest.devDependencies) {
    if (isPathDependency(dependency)) {
      dependency.path = normalizePath(dependency.path, relativeTo);
    }
  }

  if (manifest.target) {
    manifest.target.path = normalizePath(manifest.target.path, relativeTo);
  }

  return manifest;
}

function normalizePath(path: string, relativeTo: string = FIXTURES): string {
  return relative(relativeTo, path);
}
