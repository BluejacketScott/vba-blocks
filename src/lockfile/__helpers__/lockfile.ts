export function normalizeLockfile(lockfile: string): string {
	return lockfile.replace(/vba\-blocks v\d.\d.\d/g, "vba-blocks v#.#.#");
}
