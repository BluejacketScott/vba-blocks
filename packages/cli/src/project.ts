import { join } from 'path';
import env from './env';
import { Config, loadConfig } from './config';
import { Manifest, Source, Reference, loadManifest } from './manifest';
import { Workspace } from './workspace';
import SourceManager, { Registration } from './sources';
import resolve, { DependencyGraph, getRegistration } from './resolve';

import { parallel } from './utils';

export interface Project {
  manifest: Manifest;
  workspace: Workspace;
  packages: DependencyGraph;

  config: Config;
  paths: {
    root: string;
    dir: string;
    build: string;
    backup: string;
  };
}

export async function loadProject(dir?: string): Promise<Project> {
  const manifest = await loadManifest(dir || env.cwd);

  // TODO Load workspace
  const workspace: Workspace = {
    root: manifest,
    members: []
  };

  const config = await loadConfig(manifest, workspace);
  const packages = await resolve(config, workspace);

  const paths = {
    root: workspace.root.dir,
    dir: manifest.dir,
    build: join(manifest.dir, 'build'),
    backup: join(manifest.dir, 'build', '.backup')
  };

  return {
    manifest,
    workspace,
    packages,
    config,
    paths
  };
}

export async function fetchDependencies(project: Project): Promise<Manifest[]> {
  const manager = new SourceManager(project.config);
  const manifests = await parallel(
    project.manifest.dependencies,
    async dependency => {
      const registration = getRegistration(project.packages, dependency);
      const path = await manager.fetch(registration);
      const manifest = await loadManifest(path);

      return manifest;
    }
  );

  return manifests;
}
