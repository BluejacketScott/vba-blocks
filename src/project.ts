import env from './env';
import { join, normalize } from './utils/path';
import { tmpFolder } from './utils/fs';
import parallel from './utils/parallel';
import { Config, loadConfig } from './config';
import { Manifest, loadManifest } from './manifest';
import { Workspace, loadWorkspace } from './workspace';
import { fetch } from './sources';
import resolve, { DependencyGraph } from './resolve';
import { readLockfile, isLockfileValid } from './lockfile';

const debug = require('debug')('vba-blocks:project');

export interface Project {
  manifest: Manifest;
  workspace: Workspace;
  packages: DependencyGraph;

  config: Config;
  paths: {
    root: string;
    dir: string;
    build: string;
    backup: string;
    staging: string;
  };
  has_dirty_lockfile: boolean;
}

/**
 * Load project from given directory / cwd
 *
 * 1. Load manifest in dir
 * 2. Load workspace from manifest
 * 3. Load lockfile
 * 4. Resolve dependencies
 */
export async function loadProject(dir: string = env.cwd): Promise<Project> {
  const manifest = await loadManifest(dir);

  const config = await loadConfig();
  const workspace = await loadWorkspace(manifest);
  const lockfile = await readLockfile(workspace.root.dir);

  // Resolve packages from lockfile or from sources
  const has_dirty_lockfile = !lockfile || !isLockfileValid(lockfile, workspace);
  debug(
    !has_dirty_lockfile
      ? 'Loading packages from lockfile'
      : 'Resolving packages'
  );

  const packages = !has_dirty_lockfile
    ? lockfile!.packages
    : await resolve(config, workspace, lockfile ? lockfile.packages : []);

  const paths = {
    root: workspace.root.dir,
    dir: manifest.dir,
    build: join(manifest.dir, 'build'),
    backup: join(manifest.dir, 'build', '.backup'),
    staging: await tmpFolder({ dir: env.staging })
  };

  return {
    manifest,
    workspace,
    packages,
    config,
    paths,
    has_dirty_lockfile
  };
}

export interface FetchProject {
  manifest: Manifest;
  packages: DependencyGraph;
  config: Config;
}

/**
 * Fetch all dependencies for project
 * (based on already resolved project.packages)
 *
 * After sources complete fetches, manifests are loaded and returned for each package
 */
export async function fetchDependencies(
  project: FetchProject
): Promise<Manifest[]> {
  const manifests = await parallel(
    project.packages,
    async registration => {
      const path = await fetch(project.config.sources, registration);
      const manifest = await loadManifest(path);

      return manifest;
    },
    { progress: env.reporter.progress('Fetch dependencies') }
  );

  return manifests;
}

export interface ProjectOptions {
  type?: 'project' | 'package';
}

/**
 * Initialize new project
 *
 * Minimum requirements: name and dir
 */
export async function initProject(
  name: string,
  dir: string,
  options: ProjectOptions = {}
): Promise<Project> {
  dir = normalize(dir);
  const { type = 'project' } = options;

  const config = await loadConfig();

  // TODO load defaults from config
  const version = '0.1.0';
  const authors: string[] = [];
  const license = 'MIT';

  // Manually generate manifest and project
  // (may be included in project in the future)
  const manifest: Manifest = {
    type,
    name,
    version,
    metadata: {
      authors,
      publish: type === 'package',
      license
    },
    dependencies: [],
    src: [],
    references: [],
    targets: [],
    dir
  };

  const workspace = await loadWorkspace(manifest);

  const project: Project = {
    manifest,
    workspace,
    packages: [],
    config,
    paths: {
      root: workspace.root.dir,
      dir,
      build: join(dir, 'build'),
      backup: join(dir, 'build', '.backup'),
      staging: await tmpFolder({ dir: env.staging })
    },
    has_dirty_lockfile: true
  };

  return project;
}
