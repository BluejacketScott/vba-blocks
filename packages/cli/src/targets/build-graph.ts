import { Source, Reference } from '../manifest';
import { Project, fetchDependencies, loadManifests } from '../project';
import { BuildOptions } from './build-target';

export interface BuildGraph {
  src: Source[];
  references: Reference[];
}

export async function createBuildGraph(
  project: Project,
  options: BuildOptions
): Promise<BuildGraph> {
  const src: Map<string, Source> = new Map();
  const references: Map<string, Reference> = new Map();

  const manifests = project.manifests || (await loadManifests(project));

  for (const manifest of manifests) {
    for (const source of manifest.src) {
      const { name } = source;
      if (src.has(name))
        throw new Error(`Conflicting source named "${source.name}"`);
      src.set(name, source);
    }
    for (const reference of manifest.references) {
      const { name, guid } = reference;
      if (references.has(name) && references.get(name)!.guid !== guid)
        throw new Error(`Conficting reference named "${reference.name}"`);
      references.set(name, reference);
    }
  }

  return { src: [...src.values()], references: [...references.values()] };
}
