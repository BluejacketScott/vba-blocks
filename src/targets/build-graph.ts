import { Manifest, Source, Reference } from '../manifest';
import { Project } from '../project';
import { BuildOptions } from './build-target';

export interface BuildGraph {
  src: Source[];
  references: Reference[];
}

export async function createBuildGraph(
  project: Project,
  dependencies: Manifest[],
  options: BuildOptions
): Promise<BuildGraph> {
  const src: Map<string, Source> = new Map();
  const references: Map<string, Reference> = new Map();
  const manifests = [project.manifest, ...dependencies];

  for (const manifest of manifests) {
    for (const source of manifest.src) {
      const { name } = source;
      if (src.has(name))
        throw new Error(`Conflicting source named "${source.name}"`);
      src.set(name, source);
    }
    for (const reference of manifest.references) {
      const { name, guid } = reference;
      if (references.has(name) && references.get(name)!.guid !== guid)
        throw new Error(`Conficting reference named "${reference.name}"`);
      references.set(name, reference);
    }
  }

  return { src: [...src.values()], references: [...references.values()] };
}
