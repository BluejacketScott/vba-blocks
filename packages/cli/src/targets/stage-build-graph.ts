import { basename } from 'path';
import { BuildGraph } from './build-graph';
import { getStaging, ensureDir, copyFile, parallel, unixJoin } from '../utils';
import env from '../env';

// To avoid "Grant File Access" prompts on Mac,
// files to be imported/exported from VBA are staged to
// a location that can freely share files with Office:
// "HOME/Library/Group Containers/##########.Office"
//
// See: http://www.rondebruin.nl/mac/mac034.htm

export default async function stageBuildGraph(
  graph: BuildGraph
): Promise<BuildGraph> {
  if (env.isWindows) return graph;

  const staging = env.staging || (env.staging = await getStaging());

  const src = await parallel(
    graph.src,
    async source => {
      const { name, path, optional } = source;

      const dest = unixJoin(staging, basename(source.path));
      await copyFile(source.path, dest);

      return { name, path: dest, optional, original: path };
    },
    { progress: env.reporter.progress('Staging dependencies') }
  );

  return Object.assign({}, graph, { src });
}
