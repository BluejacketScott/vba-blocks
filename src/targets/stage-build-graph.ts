import { basename } from 'path';
import { Project } from '../project';
import { BuildGraph } from './build-graph';
import { ensureDir, copyFile, parallel, unixJoin, emptyDir } from '../utils';
import env from '../env';

// To avoid "Grant File Access" prompts on Mac,
// files to be imported/exported from VBA are staged to
// a location that can freely share files with Office:
// "HOME/Library/Group Containers/##########.Office"
//
// See: http://www.rondebruin.nl/mac/mac034.htm

export default async function stageBuildGraph(
  project: Project,
  graph: BuildGraph
): Promise<{ staged: string | null; graph: BuildGraph }> {
  if (env.isWindows) return { staged: null, graph };

  const staged = unixJoin(project.paths.staging, 'import');
  await ensureDir(staged);
  await emptyDir(staged);

  const src = await parallel(
    graph.src,
    async source => {
      const { name, path, optional } = source;

      const dest = unixJoin(staged, basename(source.path));
      await copyFile(source.path, dest);

      return { name, path: dest, optional, original: path };
    },
    { progress: env.reporter.progress('Staging dependencies') }
  );

  return { staged, graph: { ...graph, src } };
}
