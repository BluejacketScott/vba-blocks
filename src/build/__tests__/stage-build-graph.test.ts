import { loadBuildGraph } from '../build-graph';
import { setup, reset } from '../../../tests/__helpers__/project';
import { standard } from '../../../tests/__fixtures__';

import { stageBuildGraph, ImportGraph } from '../stage-build-graph';

afterEach(reset);

test('should stage build graph', async () => {
  const { project, dependencies } = await setup(standard);
  const staging = project.paths.staging;

  const graph = await loadBuildGraph(project, dependencies);
  const import_graph = await stageBuildGraph(graph, staging);

  expect(normalizeGraph(import_graph, staging)).toMatchSnapshot();
});

function normalizeGraph(graph: ImportGraph, staging: string): ImportGraph {
  const { name, components, references } = graph;

  return {
    name,
    components: components.map(component => {
      const { name, path } = component;
      return { name, path: path.replace(staging, 'staging') };
    }),
    references
  };
}
