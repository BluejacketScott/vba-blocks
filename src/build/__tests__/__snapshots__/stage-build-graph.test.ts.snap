// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should stage build graph for Mac 1`] = `
Object {
  "components": Array [
    Object {
      "name": "a",
      "path": "staging\\\\a.bas",
    },
    Object {
      "name": "d",
      "path": "staging\\\\d.cls",
    },
    Object {
      "name": "f",
      "path": "staging\\\\f.bas",
    },
    Object {
      "name": "g",
      "path": "staging\\\\g.cls",
    },
  ],
  "name": "VBAProject",
  "references": Array [],
}
`;
